File: modules/wem-admin-ui/src/main/resources/web/admin/common/js/query/expr/ValueExpr.ts
module api.query.expr {

    export class ValueExpr implements Expression {

        private value: api.data.Value;

        constructor(value: api.data.Value) {
            this.value = value;
        }

        getValue(): api.data.Value {
            return this.value;
        }

        static stringValue(value: string): ValueExpr {
            return new ValueExpr(new api.data.Value(value, api.data.type.ValueTypes.STRING));
        }

        toString() {
            var type: api.data.type.ValueType = this.value.getType();

            if (type == api.data.type.ValueTypes.DOUBLE) {
                return this.value.asString();
            }

            if (type == api.data.type.ValueTypes.DATE_TIME) {
                return this.typecastFunction("dateTime", this.value.asString());
            }

            if (type == api.data.type.ValueTypes.GEO_POINT) {
                return this.typecastFunction("geoPoint", this.value.asString());
            }

            return this.quoteString(this.value.asString());
        }

        private typecastFunction(name: string, argument: string): string {
            return name + "(" + this.quoteString(argument) + ")";
        }

        private quoteString(value: string): string {
            if (value.indexOf("'") > -1) {
                return "\"" + value + "\"";
            }
            else {
                return "'" + value + "'";
            }
        }

        public static string(value: string): ValueExpr {
            return new ValueExpr(new api.data.Value(value, api.data.type.ValueTypes.STRING));
        }

        public static number(value: Number): ValueExpr {
            return new ValueExpr(new api.data.Value(value, api.data.type.ValueTypes.DOUBLE));
        }

        public static dateTime(value: Date): ValueExpr {
            return new ValueExpr(new api.data.Value(value, api.data.type.ValueTypes.DATE_TIME));
        }

        public static geoPoint(value: string): ValueExpr {
            return new ValueExpr(new api.data.Value(value, api.data.type.ValueTypes.GEO_POINT));
        }
    }
}

File: modules/core/core-api/src/main/java/com/enonic/xp/query/expr/ValueExpr.java
package com.enonic.xp.query.expr;

import java.time.Instant;
import java.time.format.DateTimeFormatter;

import com.enonic.xp.annotation.PublicApi;
import com.enonic.xp.data.Value;
import com.enonic.xp.data.ValueFactory;
import com.enonic.xp.data.ValueType;
import com.enonic.xp.data.ValueTypes;

@PublicApi
public final class ValueExpr
    implements Expression
{
    private final Value value;

    private ValueExpr( final Value value )
    {
        this.value = value;
    }

    public static ValueExpr string( final String value )
    {
        return new ValueExpr( ValueFactory.newString( value ) );
    }

    public static ValueExpr number( final Number value )
    {
        return new ValueExpr( ValueFactory.newDouble( value.doubleValue() ) );
    }

    public static ValueExpr instant( final String value )
    {
        return new ValueExpr( ValueFactory.newDateTime( ValueTypes.DATE_TIME.convert( value ) ) );
    }

    public static ValueExpr dateTime( final String value )
    {
        return new ValueExpr( ValueFactory.newDateTime( Instant.from( DateTimeFormatter.ISO_DATE_TIME.parse( value ) ) ) );
    }

    public static ValueExpr localDateTime( final String value )
    {
        return new ValueExpr( ValueFactory.newLocalDateTime( ValueTypes.LOCAL_DATE_TIME.convert( value ) ) );
    }

    public static ValueExpr time( final String value )
    {
        return new ValueExpr( ValueFactory.newLocalTime( ValueTypes.LOCAL_TIME.convert( value ) ) );
    }

    public static ValueExpr date( final String value )
    {
        return new ValueExpr( ValueFactory.newLocalDate( ValueTypes.LOCAL_DATE.convert( value ) ) );
    }

    public static ValueExpr geoPoint( final String value )
    {
        return new ValueExpr( ValueFactory.newGeoPoint( ValueTypes.GEO_POINT.convert( value ) ) );
    }

    public Value getValue()
    {
        return this.value;
    }

    @Override
    public String toString()
    {
        final ValueType type = this.value.getType();
        if ( type.equals( ValueTypes.DOUBLE ) )
        {
            return this.value.asString();
        }

        if ( type.equals( ValueTypes.DATE_TIME ) )
        {
            return typecastFunction( "instant", this.value.asString() );
        }

        if ( type.equals( ValueTypes.GEO_POINT ) )
        {
            return typecastFunction( "geoPoint", this.value.asString() );
        }

        return quoteString( this.value.asString() );
    }

    private String typecastFunction( final String name, final String argument )
    {
        return name + "(" + quoteString( argument ) + ")";
    }

    private String quoteString( final String value )
    {
        if ( value.contains( "'" ) )
        {
            return "\"" + value + "\"";
        }
        else
        {
            return "'" + value + "'";
        }
    }
}
